# SQL for Data analysis

##### We are going to work with Titanic dataset using SQL within Python. SQL (Structured Query Lanuage) is used to manipulate structured data. SQL is used to access data stored in databases. There are many SQL dialects. These dialects mostly have common syntax except some which differ here and there. The syntax used in this project are widely used in Companies.

Let's start with installing modules which we will be using in this project. pandasql modle will help us to run SQL on python.

In [ ]:
!pip install pandas
!pip install pandasql

In [3]:
import pandas as pd
import pandasql as ps

This is the famous dataset used for ananlysis purpose. In 1912 there was an unfortunate event of Titanic disaster. It will give us the details of passenger traveling on Titanic ship and if they were able to survive by luck. We are importing and having an overview of the data.

In [4]:
titanic_disaster_dataset = pd.read_csv("https://raw.githubusercontent.com/sumeetcode88/Data-Analysis-with-SQL/main/titanic.csv")
print(titanic_disaster_dataset.shape, titanic_disaster_dataset[['PassengerId']].drop_duplicates().shape)
titanic_disaster_dataset.head()

(891, 12) (891, 1)


PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In SQL we can pull whole data from dataset using "*"

In [8]:
ps.sqldf("""

SELECT * FROM titanic_disaster_dataset

""")

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500  None        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250  None        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500  None        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000  None        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500  None        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500  None        Q  

[891 rows x 12 columns]

We use "AS" keyword to give alias for a table or column. 
It is a good practice to do and easy to understand, specially when we use JOINS. Also, it helps to shorten the name of data if it has long names. We have also used LIMIT to limit the view of data. If you see the above output you will see this dataset has 891 rows so instead of pulling all rows we  have used LIMIT for a sample view.

In [9]:
ps.sqldf("""

SELECT df.* FROM titanic_disaster_dataset AS df
LIMIT 5

""")

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500  None        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250  None        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500  None        S

If we wish to see some specific columns drom dataset we can do it as below.

In [10]:
ps.sqldf("""

SELECT df.PassengerId, df.Name
FROM titanic_disaster_dataset AS df
LIMIT 5

""")

PassengerId                                               Name
0            1                            Braund, Mr. Owen Harris
1            2  Cumings, Mrs. John Bradley (Florence Briggs Th...
2            3                             Heikkinen, Miss. Laina
3            4       Futrelle, Mrs. Jacques Heath (Lily May Peel)
4            5                           Allen, Mr. William Henry

Instead of actual column names under dataset we have given different names to column using Alias.

In [11]:
ps.sqldf("""

SELECT df.PassengerId AS Ticket_Number, df.Name AS Passenger_Name
FROM titanic_disaster_dataset AS df
LIMIT 5

""")

Ticket_Number                                     Passenger_Name
0              1                            Braund, Mr. Owen Harris
1              2  Cumings, Mrs. John Bradley (Florence Briggs Th...
2              3                             Heikkinen, Miss. Laina
3              4       Futrelle, Mrs. Jacques Heath (Lily May Peel)
4              5                           Allen, Mr. William Henry

### WHERE Clause


WHERE clause used to pull rows based on condition given.
Here we are asking SQL to pull rows where Passenger age is less than 15

In [12]:
ps.sqldf("""

SELECT df.*
FROM titanic_disaster_dataset AS df
WHERE df.age < 15

""")

PassengerId  Survived  Pclass                                     Name  \
0             8         0       3           Palsson, Master. Gosta Leonard   
1            10         1       2      Nasser, Mrs. Nicholas (Adele Achem)   
2            11         1       3          Sandstrom, Miss. Marguerite Rut   
3            15         0       3     Vestrom, Miss. Hulda Amanda Adolfina   
4            17         0       3                     Rice, Master. Eugene   
..          ...       ...     ...                                      ...   
73          828         1       2                    Mallet, Master. Andre   
74          832         1       2          Richards, Master. George Sibley   
75          851         0       3  Andersson, Master. Sigvard Harald Elias   
76          853         0       3                  Boulos, Miss. Nourelain   
77          870         1       3          Johnson, Master. Harold Theodor   

       Sex    Age  SibSp  Parch           Ticket     Fare Cabin Embarked  
0     male   2.00      3      1           349909  21.0750  None        S  
1   female  14.00      1      0           237736  30.0708  None        C  
2   female   4.00      1      1          PP 9549  16.7000    G6        S  
3   female  14.00      0      0           350406   7.8542  None        S  
4     male   2.00      4      1           382652  29.1250  None        Q  
..     ...    ...    ...    ...              ...      ...   ...      ...  
73    male   1.00      0      2  S.C./PARIS 2079  37.0042  None        C  
74    male   0.83      1      1            29106  18.7500  None        S  
75    male   4.00      4      2           347082  31.2750  None        S  
76  female   9.00      1      1             2678  15.2458  None        C  
77    male   4.00      1      1           347742  11.1333  None        S  

[78 rows x 12 columns]

Sometimes the data we recieve has missing values and we need to treat these values for further analysis. In SQL we use IS NULL keyword to find Null values and NOT NULL for Non-Null values.

In [13]:
ps.sqldf("""

SELECT df.*
FROM titanic_disaster_dataset AS df
WHERE df.age IS NULL

""")

PassengerId  Survived  Pclass                                      Name  \
0              6         0       3                          Moran, Mr. James   
1             18         1       2              Williams, Mr. Charles Eugene   
2             20         1       3                   Masselmani, Mrs. Fatima   
3             27         0       3                   Emir, Mr. Farred Chehab   
4             29         1       3             O'Dwyer, Miss. Ellen "Nellie"   
..           ...       ...     ...                                       ...   
172          860         0       3                          Razi, Mr. Raihed   
173          864         0       3         Sage, Miss. Dorothy Edith "Dolly"   
174          869         0       3               van Melkebeke, Mr. Philemon   
175          879         0       3                        Laleff, Mr. Kristo   
176          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   

        Sex   Age  SibSp  Parch      Ticket     Fare Cabin Embarked  
0      male  None      0      0      330877   8.4583  None        Q  
1      male  None      0      0      244373  13.0000  None        S  
2    female  None      0      0        2649   7.2250  None        C  
3      male  None      0      0        2631   7.2250  None        C  
4    female  None      0      0      330959   7.8792  None        Q  
..      ...   ...    ...    ...         ...      ...   ...      ...  
172    male  None      0      0        2629   7.2292  None        C  
173  female  None      8      2    CA. 2343  69.5500  None        S  
174    male  None      0      0      345777   9.5000  None        S  
175    male  None      0      0      349217   7.8958  None        S  
176  female  None      1      2  W./C. 6607  23.4500  None        S  

[177 rows x 12 columns]

In [14]:
ps.sqldf("""

SELECT df.*
FROM titanic_disaster_dataset AS df
WHERE df.age NOT NULL

""")

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
709          886         0       3   
710          887         0       2   
711          888         1       1   
712          890         1       1   
713          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
709               Rice, Mrs. William (Margaret Norton)  female  39.0      0   
710                              Montvila, Rev. Juozas    male  27.0      0   
711                       Graham, Miss. Margaret Edith  female  19.0      0   
712                              Behr, Mr. Karl Howell    male  26.0      0   
713                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500  None        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250  None        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500  None        S  
..     ...               ...      ...   ...      ...  
709      5            382652  29.1250  None        Q  
710      0            211536  13.0000  None        S  
711      0            112053  30.0000   B42        S  
712      0            111369  30.0000  C148        C  
713      0            370376   7.7500  None        Q  

[714 rows x 12 columns]

Use BETWEEN to filter numerical values of rows. Here we have pull all passengers between the age of 5 and 10

In [16]:
ps.sqldf("""

SELECT 
    df.*
FROM titanic_disaster_dataset AS df
WHERE df.age BETWEEN 5 AND 10

""")

PassengerId  Survived  Pclass  \
0            25         0       3   
1            51         0       3   
2            59         1       2   
3           148         0       3   
4           166         1       3   
5           183         0       3   
6           234         1       3   
7           238         1       2   
8           279         0       3   
9           420         0       3   
10          449         1       3   
11          481         0       3   
12          490         1       3   
13          536         1       2   
14          542         0       3   
15          550         1       2   
16          635         0       3   
17          721         1       2   
18          752         1       3   
19          778         1       3   
20          788         0       3   
21          814         0       3   
22          820         0       3   
23          853         0       3   

                                               Name     Sex   Age  SibSp  \
0                     Palsson, Miss. Torborg Danira  female   8.0      3   
1                        Panula, Master. Juha Niilo    male   7.0      4   
2                      West, Miss. Constance Mirium  female   5.0      1   
3                  Ford, Miss. Robina Maggie "Ruby"  female   9.0      2   
4   Goldsmith, Master. Frank John William "Frankie"    male   9.0      0   
5             Asplund, Master. Clarence Gustaf Hugo    male   9.0      4   
6                    Asplund, Miss. Lillian Gertrud  female   5.0      4   
7                  Collyer, Miss. Marjorie "Lottie"  female   8.0      0   
8                                Rice, Master. Eric    male   7.0      4   
9                         Van Impe, Miss. Catharina  female  10.0      0   
10                   Baclini, Miss. Marie Catherine  female   5.0      2   
11                   Goodwin, Master. Harold Victor    male   9.0      5   
12            Coutts, Master. Eden Leslie "Neville"    male   9.0      1   
13                           Hart, Miss. Eva Miriam  female   7.0      0   
14             Andersson, Miss. Ingeborg Constanzia  female   9.0      4   
15                   Davies, Master. John Morgan Jr    male   8.0      1   
16                               Skoog, Miss. Mabel  female   9.0      3   
17                Harper, Miss. Annie Jessie "Nina"  female   6.0      0   
18                              Moor, Master. Meier    male   6.0      0   
19                    Emanuel, Miss. Virginia Ethel  female   5.0      0   
20                        Rice, Master. George Hugh    male   8.0      4   
21               Andersson, Miss. Ebba Iris Alfrida  female   6.0      4   
22                     Skoog, Master. Karl Thorsten    male  10.0      3   
23                          Boulos, Miss. Nourelain  female   9.0      1   

    Parch        Ticket     Fare Cabin Embarked  
0       1        349909  21.0750  None        S  
1       1       3101295  39.6875  None        S  
2       2    C.A. 34651  27.7500  None        S  
3       2    W./C. 6608  34.3750  None        S  
4       2        363291  20.5250  None        S  
5       2        347077  31.3875  None        S  
6       2        347077  31.3875  None        S  
7       2    C.A. 31921  26.2500  None        S  
8       1        382652  29.1250  None        Q  
9       2        345773  24.1500  None        S  
10      1          2666  19.2583  None        C  
11      2       CA 2144  46.9000  None        S  
12      1    C.A. 37671  15.9000  None        S  
13      2  F.C.C. 13529  26.2500  None        S  
14      2        347082  31.2750  None        S  
15      1    C.A. 33112  36.7500  None        S  
16      2        347088  27.9000  None        S  
17      1        248727  33.0000  None        S  
18      1        392096  12.4750  E121        S  
19      0        364516  12.4750  None        S  
20      1        382652  29.1250  None        Q  
21      2        347082  31.2750  None        S  
22      2        347088  27

With WHERE clause if multiple conditions are to be passed we use "AND" or "OR" or combination of these two.

In [17]:
ps.sqldf("""

SELECT df.*
FROM titanic_disaster_dataset AS df
WHERE df.Sex = "female" AND Survived = 1
    
""")

PassengerId  Survived  Pclass  \
0              2         1       1   
1              3         1       3   
2              4         1       1   
3              9         1       3   
4             10         1       2   
..           ...       ...     ...   
228          875         1       2   
229          876         1       3   
230          880         1       1   
231          881         1       2   
232          888         1       1   

                                                  Name     Sex   Age  SibSp  \
0    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
1                               Heikkinen, Miss. Laina  female  26.0      0   
2         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
3    Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
4                  Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
..                                                 ...     ...   ...    ...   
228              Abelson, Mrs. Samuel (Hannah Wizosky)  female  28.0      1   
229                   Najib, Miss. Adele Kiamie "Jane"  female  15.0      0   
230      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   
231       Shelley, Mrs. William (Imanita Parrish Hall)  female  25.0      0   
232                       Graham, Miss. Margaret Edith  female  19.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0          PC 17599  71.2833   C85        C  
1        0  STON/O2. 3101282   7.9250  None        S  
2        0            113803  53.1000  C123        S  
3        2            347742  11.1333  None        S  
4        0            237736  30.0708  None        C  
..     ...               ...      ...   ...      ...  
228      0         P/PP 3381  24.0000  None        C  
229      0              2667   7.2250  None        C  
230      1             11767  83.1583   C50        C  
231      1            230433  26.0000  None        S  
232      0            112053  30.0000   B42        S  

[233 rows x 12 columns]

We saw total 233 females survived in this disaster. Lets see one more example of condition based statement. IT has "AND" "OR" together and we need to use paranthesis to make understand SQL which condition should be checked together

In [18]:
ps.sqldf("""

SELECT df.*
FROM titanic_disaster_dataset AS df
WHERE df.Pclass=2 OR (df.Fare < 20 AND df.Cabin IS NULL)
    
""")

PassengerId  Survived  Pclass  \
0              1         0       3   
1              3         1       3   
2              5         0       3   
3              6         0       3   
4              9         1       3   
..           ...       ...     ...   
575          883         0       3   
576          884         0       2   
577          885         0       3   
578          887         0       2   
579          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1                               Heikkinen, Miss. Laina  female  26.0      0   
2                             Allen, Mr. William Henry    male  35.0      0   
3                                     Moran, Mr. James    male   NaN      0   
4    Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
..                                                 ...     ...   ...    ...   
575                       Dahlberg, Miss. Gerda Ulrika  female  22.0      0   
576                      Banfield, Mr. Frederick James    male  28.0      0   
577                             Sutehall, Mr. Henry Jr    male  25.0      0   
578                              Montvila, Rev. Juozas    male  27.0      0   
579                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500  None        S  
1        0  STON/O2. 3101282   7.9250  None        S  
2        0            373450   8.0500  None        S  
3        0            330877   8.4583  None        Q  
4        2            347742  11.1333  None        S  
..     ...               ...      ...   ...      ...  
575      0              7552  10.5167  None        S  
576      0  C.A./SOTON 34068  10.5000  None        S  
577      0   SOTON/OQ 392076   7.0500  None        S  
578      0            211536  13.0000  None        S  
579      0            370376   7.7500  None        Q  

[580 rows x 12 columns]

We use DISTINCT for retrieve unique values. Let us see unique values in Cabin column. 

In [19]:
ps.sqldf("""

SELECT DISTINCT df.Cabin
FROM titanic_disaster_dataset AS df    
""")

Cabin
0    None
1     C85
2    C123
3     E46
4      G6
..    ...
143   E17
144   A24
145   C50
146   B42
147  C148

[148 rows x 1 columns]

We found 148 unique values under Cabin Column. Now lets filter rows on the basis of multiple values under Cabin column. "IN" is used to filter rows with multiple values.

In [20]:
ps.sqldf("""

SELECT df.*
FROM titanic_disaster_dataset AS df
WHERE df.Cabin IN ('E17', 'C85', 'G6')
    
""")

PassengerId  Survived  Pclass  \
0            2         1       1   
1           11         1       3   
2          206         0       3   
3          252         0       3   
4          395         1       3   
5          858         1       1   

                                                Name     Sex   Age  SibSp  \
0  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
1                    Sandstrom, Miss. Marguerite Rut  female   4.0      1   
2                         Strom, Miss. Telma Matilda  female   2.0      0   
3         Strom, Mrs. Wilhelm (Elna Matilda Persson)  female  29.0      1   
4  Sandstrom, Mrs. Hjalmar (Agnes Charlotta Bengt...  female  24.0      0   
5                             Daly, Mr. Peter Denis     male  51.0      0   

   Parch    Ticket     Fare Cabin Embarked  
0      0  PC 17599  71.2833   C85        C  
1      1   PP 9549  16.7000    G6        S  
2      1    347054  10.4625    G6        S  
3      1    347054  10.4625    G6        S  
4      2   PP 9549  16.7000    G6        S  
5      0    113055  26.5500   E17        S

LIKE keywrod is for columns having String/Text. Suppose we want to find names of passengers starting with "A" or having title as "MRS" or name ending with "li", we can use LIKE to pull rows having these substring from string columns. Note that it is not case sensitive.

In [32]:
ps.sqldf("""

SELECT df.*
FROM titanic_disaster_dataset AS df
WHERE df.Name LIKE "%MRS%"
LIMIT 5

""")

PassengerId  Survived  Pclass  \
0            2         1       1   
1            4         1       1   
2            9         1       3   
3           10         1       2   
4           16         1       2   

                                                Name     Sex   Age  SibSp  \
0  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
1       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
2  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
3                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
4                   Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0   

   Parch    Ticket     Fare Cabin Embarked  
0      0  PC 17599  71.2833   C85        C  
1      0    113803  53.1000  C123        S  
2      2    347742  11.1333  None        S  
3      0    237736  30.0708  None        C  
4      0    248706  16.0000  None        S

In [34]:
ps.sqldf("""

SELECT df.*
FROM titanic_disaster_dataset AS df
WHERE df.Embarked LIKE "S"
LIMIT 5

""")

PassengerId  Survived  Pclass  \
0            1         0       3   
1            3         1       3   
2            4         1       1   
3            5         0       3   
4            7         0       1   

                                           Name     Sex   Age  SibSp  Parch  \
0                       Braund, Mr. Owen Harris    male  22.0      1      0   
1                        Heikkinen, Miss. Laina  female  26.0      0      0   
2  Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1      0   
3                      Allen, Mr. William Henry    male  35.0      0      0   
4                       McCarthy, Mr. Timothy J    male  54.0      0      0   

             Ticket     Fare Cabin Embarked  
0         A/5 21171   7.2500  None        S  
1  STON/O2. 3101282   7.9250  None        S  
2            113803  53.1000  C123        S  
3            373450   8.0500  None        S  
4             17463  51.8625   E46        S

In [36]:
ps.sqldf("""

SELECT df.*
FROM titanic_disaster_dataset AS df
WHERE df.Name LIKE "%IS"
LIMIT 5

""")

PassengerId  Survived  Pclass                            Name   Sex   Age  \
0            1         0       3         Braund, Mr. Owen Harris  male  22.0   
1           47         0       3               Lennon, Mr. Denis  male   NaN   
2           88         0       3   Slocovski, Mr. Selman Francis  male   NaN   
3          221         1       3  Sunderland, Mr. Victor Francis  male  16.0   
4          457         0       1       Millet, Mr. Francis Davis  male  65.0   

   SibSp  Parch           Ticket   Fare Cabin Embarked  
0      1      0        A/5 21171   7.25  None        S  
1      1      0           370371  15.50  None        Q  
2      0      0  SOTON/OQ 392086   8.05  None        S  
3      0      0  SOTON/OQ 392089   8.05  None        S  
4      0      0            13509  26.55   E38        S

We can create new column from existing columns. Lets look at below example, we are performing calculations on Age column and out put is shown as a new column which is not actually in dataframe.

In [39]:
ps.sqldf("""

SELECT 
    df.Age AS Actual_Age,
    df.Age * df.Age AS Age_square,
    df.Age + 1 AS Age_Plus_One
FROM titanic_disaster_dataset AS df
LIMIT 5    
    
""")

Actual_Age  Age_square  Age_Plus_One
0        22.0       484.0          23.0
1        38.0      1444.0          39.0
2        26.0       676.0          27.0
3        35.0      1225.0          36.0
4        35.0      1225.0          36.0

Below statement performs operation on String columns. We have concatinated two columns using double pipe ||

In [42]:
ps.sqldf("""

SELECT 
    df.Sex,
    df.Ticket,
    df.Sex || "-" || df.Ticket AS Gender_Plus_ticket
FROM titanic_disaster_dataset AS df
LIMIT 5    
    
""")

Sex            Ticket       Gender_Plus_ticket
0    male         A/5 21171           male-A/5 21171
1  female          PC 17599          female-PC 17599
2  female  STON/O2. 3101282  female-STON/O2. 3101282
3  female            113803            female-113803
4    male            373450              male-373450

"CASE WHEN" is used for creating a new column based on conditions we pass. In our data have AGE column having differnt age groups, let us check how we can create a column for it. We have used IN to make sure if age group are defined correctly.

In [67]:
ps.sqldf("""

SELECT 
    df.Name, df.PassengerID, df.Age,
    CASE WHEN df.Age IS NULL THEN 'Missing'
        WHEN df.Age <12 THEN 'Kid'
        WHEN df.Age <20 THEN 'Teenage'
        WHEN df.Age <60 THEN 'Adult'
    ELSE 'Senior Citizen' END AS Age_Group
FROM titanic_disaster_dataset AS df
WHERE df.Age IN (5,15,25,65)
LIMIT 10   
    
""")

Name  PassengerId   Age       Age_Group
0            McGowan, Miss. Anna "Annie"           23  15.0         Teenage
1         Ostby, Mr. Engelhart Cornelius           55  65.0  Senior Citizen
2           West, Miss. Constance Mirium           59   5.0             Kid
3                Moen, Mr. Sigurd Hansen           76  25.0           Adult
4         Sobey, Mr. Samuel James Hayden          135  25.0           Adult
5         Asplund, Miss. Lillian Gertrud          234   5.0             Kid
6  Lindahl, Miss. Agda Thorilda Viktoria          247  25.0           Adult
7               Persson, Mr. Ernst Ulrik          268  25.0           Adult
8           Tornquist, Mr. William Henry          272  25.0           Adult
9                       Duane, Mr. Frank          281  65.0  Senior Citizen

SUBSTR function can be used to pull a part of string from the text columns. Below statement will pull substring from NAME column, which starts from 2nd character and go till it 4 number of characters.

In [70]:
ps.sqldf("""

SELECT df.Name, SUBSTR(df.Name,2, 4) AS Substring
FROM titanic_disaster_dataset AS df
LIMIT 5
    
""")

Name Substring
0                            Braund, Mr. Owen Harris      raun
1  Cumings, Mrs. John Bradley (Florence Briggs Th...      umin
2                             Heikkinen, Miss. Laina      eikk
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)      utre
4                           Allen, Mr. William Henry      llen

We can change case of text columns by using UPPER() and LOWER(). Refer below example.

In [74]:
ps.sqldf("""

SELECT 
    df.Name AS OG_Case_Name,
    LOWER(df.Name) AS Lcase_name,
    UPPER(df.Name) AS Ucase_name
FROM titanic_disaster_dataset AS df
LIMIT 10
    
""")

OG_Case_Name  \
0                            Braund, Mr. Owen Harris   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2                             Heikkinen, Miss. Laina   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4                           Allen, Mr. William Henry   
5                                   Moran, Mr. James   
6                            McCarthy, Mr. Timothy J   
7                     Palsson, Master. Gosta Leonard   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)   
9                Nasser, Mrs. Nicholas (Adele Achem)   

                                          Lcase_name  \
0                            braund, mr. owen harris   
1  cumings, mrs. john bradley (florence briggs th...   
2                             heikkinen, miss. laina   
3       futrelle, mrs. jacques heath (lily may peel)   
4                           allen, mr. william henry   
5                                   moran, mr. james   
6                            mccarthy, mr. timothy j   
7                     palsson, master. gosta leonard   
8  johnson, mrs. oscar w (elisabeth vilhelmina berg)   
9                nasser, mrs. nicholas (adele achem)   

                                          Ucase_name  
0                            BRAUND, MR. OWEN HARRIS  
1  CUMINGS, MRS. JOHN BRADLEY (FLORENCE BRIGGS TH...  
2                             HEIKKINEN, MISS. LAINA  
3       FUTRELLE, MRS. JACQUES HEATH (LILY MAY PEEL)  
4                           ALLEN, MR. WILLIAM HENRY  
5                                   MORAN, MR. JAMES  
6                            MCCARTHY, MR. TIMOTHY J  
7                     PALSSON, MASTER. GOSTA LEONARD  
8  JOHNSON, MRS. OSCAR W (ELISABETH VILHELMINA BERG)  
9                NASSER, MRS. NICHOLAS (ADELE ACHEM)

Here we have tried a complex query for separating first name and last name. We have used a combination of SUBSTR(), LOWER() and UPPER().

In [83]:
ps.sqldf("""

SELECT 
    df.Name AS Full_Name,    
    SUBSTR(df.Name,  instr(df.Name,".")+2, LENGTH(df.name)) AS FirstName_Ucase,
    SUBSTR(df.Name,1, instr(df.Name,",")-1) AS LastName_Ucase
FROM titanic_disaster_dataset AS df
LIMIT 5
    
""")

Full_Name  \
0                            Braund, Mr. Owen Harris   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2                             Heikkinen, Miss. Laina   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4                           Allen, Mr. William Henry   

                         FirstName_Ucase LastName_Ucase  
0                            Owen Harris         Braund  
1  John Bradley (Florence Briggs Thayer)        Cumings  
2                                  Laina      Heikkinen  
3          Jacques Heath (Lily May Peel)       Futrelle  
4                          William Henry          Allen

### AGGREGATIONS

Lets look at the aggregation of data. We have tried commonly used functions like SUM, AVG, MIN, MAX, COUNT on Age column for understanding.

In [86]:
ps.sqldf("""

SELECT 
    AVG(df.Age) AS avg_age,
    SUM(df.Age) AS tot_age,
    MIN(df.Age) AS min_age,
    MAX(df.Age) AS max_age,
    COUNT(df.PassengerId) AS tot_passengers,
    COUNT(DISTINCT df.Ticket) AS dist_tickets
FROM titanic_disaster_dataset AS df

""")

avg_age   tot_age  min_age  max_age  tot_passengers  dist_tickets
0  29.699118  21205.17     0.42     80.0             891           681

If we look at above results it did calculations, but on all rows at once. This result may not be usable. What if we do aggregation particularly based on anyone column like Embarked which has 4 unique values. So this is example of GROUP BY clause which is used to combine rows on the basis of column values i.e. we will get 4 rows and each row will be aggregated value for Emarked value.

In [90]:
ps.sqldf("""

SELECT 
    DISTINCT df.Embarked
FROM titanic_disaster_dataset AS df
    
""")

Embarked
0        S
1        C
2        Q
3     None

In [91]:
ps.sqldf("""

SELECT 
    df.Embarked,
    AVG(df.Age) AS Avg_Age,
    SUM(df.Age) AS Sum_Age,
    MIN(df.Age) AS Min_Age,
    MAX(df.Age) AS Max_Age,
    COUNT(df.PassengerId) AS Total_passengers,
    COUNT(DISTINCT df.Ticket) AS Unique_tickets
FROM titanic_disaster_dataset AS df
GROUP BY df.Embarked

""")

Embarked    Avg_Age   Sum_Age  Min_Age  Max_Age  Total_passengers  \
0     None  50.000000    100.00    38.00     62.0                 2   
1        C  30.814769   4005.92     0.42     71.0               168   
2        Q  28.089286    786.50     2.00     70.5                77   
3        S  29.445397  16312.75     0.67     80.0               644   

   Unique_tickets  
0               1  
1             122  
2              66  
3             494

In GROUP BY clause we can pass multiple columns, which will create output of unique combination of values of these multiple columns. See this example where we have also combined WHERE.

In [97]:
ps.sqldf("""

SELECT 
    df.Embarked, df.Sex,
    AVG(df.Age) AS Avg_Age,
    SUM(df.Age) AS Sum_Age,
    MIN(df.Age) AS Min_Age,
    MAX(df.Age) AS Max_Age,
    COUNT(df.PassengerId) AS Total_passengers,
    COUNT(DISTINCT df.Ticket) AS Unique_tickets
FROM titanic_disaster_dataset AS df

WHERE df.Age>18 and df.Age<60

GROUP BY df.Embarked, df.Sex

""")

Embarked     Sex    Avg_Age  Sum_Age  Min_Age  Max_Age  Total_passengers  \
0     None  female  38.000000     38.0     38.0     38.0                 1   
1        C  female  35.975000   1439.0     19.0     58.0                40   
2        C    male  33.657895   1918.5     20.0     58.0                57   
3        Q  female  28.312500    226.5     19.0     39.0                 8   
4        Q    male  33.850000    338.5     19.0     57.0                10   
5        S  female  32.803571   4592.5     19.0     58.0               140   
6        S    male  32.320819   9470.0     19.0     59.0               293   

   Unique_tickets  
0               1  
1              34  
2              54  
3               8  
4              10  
5             129  
6             279

We can also sort the data using ORDER BY. For an example lets use same query as above and add ORDER BY clause. By default it gives result in Ascending order and for Descending order we use DESC at the end.

In [100]:
ps.sqldf("""

SELECT 
    df.Embarked, df.Sex,
    AVG(df.Age) AS Avg_Age,
    SUM(df.Age) AS Sum_Age,
    MIN(df.Age) AS Min_Age,
    MAX(df.Age) AS Max_Age,
    COUNT(df.PassengerId) AS Total_passengers,
    COUNT(DISTINCT df.Ticket) AS Unique_tickets
FROM titanic_disaster_dataset AS df

WHERE df.Age>18 and df.Age<60

GROUP BY df.Embarked, df.Sex

ORDER BY Min_Age Desc
""")

Embarked     Sex    Avg_Age  Sum_Age  Min_Age  Max_Age  Total_passengers  \
0     None  female  38.000000     38.0     38.0     38.0                 1   
1        C    male  33.657895   1918.5     20.0     58.0                57   
2        C  female  35.975000   1439.0     19.0     58.0                40   
3        Q  female  28.312500    226.5     19.0     39.0                 8   
4        Q    male  33.850000    338.5     19.0     57.0                10   
5        S  female  32.803571   4592.5     19.0     58.0               140   
6        S    male  32.320819   9470.0     19.0     59.0               293   

   Unique_tickets  
0               1  
1              54  
2              34  
3               8  
4              10  
5             129  
6             279

On the aggregated information we can put filter by using HAVING clause. The difference between WHERE and HAVING is that WHERE is used on data directly from dataset and HAVING is used on aggregated data.

In [116]:
ps.sqldf("""

SELECT 
    df.Embarked, df.Sex,
    AVG(df.Age) AS Avg_Age,
    SUM(df.Age) AS Sum_Age,
    MIN(df.Age) AS Min_Age,
    MAX(df.Age) AS Max_Age,
    COUNT(df.PassengerId) AS Total_passengers,
    COUNT(DISTINCT df.Ticket) AS Unique_tickets
FROM titanic_disaster_dataset AS df

WHERE df.Age>18 and df.Age<60

GROUP BY df.Embarked, df.Sex

HAVING Avg_Age > 30

ORDER BY df.Embarked, df.Sex
""")

Embarked     Sex    Avg_Age  Sum_Age  Min_Age  Max_Age  Total_passengers  \
0     None  female  38.000000     38.0     38.0     38.0                 1   
1        C  female  35.975000   1439.0     19.0     58.0                40   
2        C    male  33.657895   1918.5     20.0     58.0                57   
3        Q    male  33.850000    338.5     19.0     57.0                10   
4        S  female  32.803571   4592.5     19.0     58.0               140   
5        S    male  32.320819   9470.0     19.0     59.0               293   

   Unique_tickets  
0               1  
1              34  
2              54  
3              10  
4             129  
5             279

## Note: Refer order of all clauses below.
## 1. "WHERE"  2. "GROUP BY"  3. "HAVING"  4. "ORDER BY"

Let us further move to JOINS. for this we will need one more table which we have created below. You will find same values in previous Emabrked column. And we have purposly missed the value Q = Queenstown for demonstrating the JOINS.

In [6]:
cities = pd.DataFrame({'code':['S', 'C', 'L'],'city':['Southampton', 'Cherbourg', 'London']})
cities

code         city
0    S  Southampton
1    C    Cherbourg
2    L       London

<b>INNER JOIN</b><br>
INNER JOIN gives rows corresponding to only the common values(in the columns used in the inner join condition) between the two tables

JOIN keyword by default is INNER JOIN. Below query shows as common values between two tables "titanic_disaster_dataset" and "cities". 
<br>"S" and "C" are picked and "Q" is missed.

In [120]:
ps.sqldf("""

SELECT 
    c.city,
    df.Embarked,
    df.Sex,
    AVG(df.Age) AS Avg_Age,
    SUM(df.Age) AS Sum_Age,
    MIN(df.Age) AS Min_Age,
    MAX(df.Age) AS Max_Age,
    COUNT(df.PassengerId) AS Total_passengers,
    COUNT(DISTINCT df.Ticket) AS Unique_tickets
    
FROM titanic_disaster_dataset AS df

INNER JOIN cities AS c
    ON df.Embarked = c.code

GROUP BY df.Embarked,
    df.Sex

""")

city Embarked     Sex    Avg_Age   Sum_Age  Min_Age  Max_Age  \
0    Cherbourg        C  female  28.344262   1729.00     0.75     60.0   
1    Cherbourg        C    male  32.998841   2276.92     0.42     71.0   
2  Southampton        S  female  27.771505   5165.50     1.00     63.0   
3  Southampton        S    male  30.291440  11147.25     0.67     80.0   

   Total_passengers  Unique_tickets  
0                73              54  
1                95              88  
2               203             158  
3               441             394

<b>LEFT JOIN</b><br>
LEFT JOIN gives all rows from the left table and only those rows from right table where the values are common in the columns used in left join condition

The table "titanic_disaster_dataset" is on LEFT and below query pulls all values of this table whereas the values not found in "cities" are showing as NULL(None).

In [121]:
ps.sqldf("""

SELECT 
    c.city,
    df.Embarked,
    df.Sex,
    AVG(df.Age) AS Avg_Age,
    SUM(df.Age) AS Sum_Age,
    MIN(df.Age) AS Min_Age,
    MAX(df.Age) AS Max_Age,
    COUNT(df.PassengerId) AS Total_passengers,
    COUNT(DISTINCT df.Ticket) AS Unique_tickets
    
FROM titanic_disaster_dataset AS df

LEFT JOIN cities AS c
    ON df.Embarked = c.code

GROUP BY df.Embarked,
    df.Sex

""")

city Embarked     Sex    Avg_Age   Sum_Age  Min_Age  Max_Age  \
0         None     None  female  50.000000    100.00    38.00     62.0   
1    Cherbourg        C  female  28.344262   1729.00     0.75     60.0   
2    Cherbourg        C    male  32.998841   2276.92     0.42     71.0   
3         None        Q  female  24.291667    291.50    15.00     39.0   
4         None        Q    male  30.937500    495.00     2.00     70.5   
5  Southampton        S  female  27.771505   5165.50     1.00     63.0   
6  Southampton        S    male  30.291440  11147.25     0.67     80.0   

   Total_passengers  Unique_tickets  
0                 2               1  
1                73              54  
2                95              88  
3                36              35  
4                41              37  
5               203             158  
6               441             394

<b>RIGHT JOIN</b><br>
RIGHT JOIN is the mirror opposite of LEFT JOIN. RIGHT JOIN gives all rows from the right table and only those rows from left table where the values are common in the columns used in right join condition.

pandasql do not support RIGHT JOIN but if we swap the tables we will get same output as RIGHT JOIN. Check below query, all values of "cities" table are pulled in the results.

In [124]:
ps.sqldf("""

SELECT 
    c.city,
    df.Embarked,
    df.Sex,
    AVG(df.Age) AS Avg_Age,
    SUM(df.Age) AS Sum_Age,
    MIN(df.Age) AS Min_Age,
    MAX(df.Age) AS Max_Age,
    COUNT(df.PassengerId) AS Total_passengers,
    COUNT(DISTINCT df.Ticket) AS Unique_tickets
    
FROM cities AS c

LEFT JOIN titanic_disaster_dataset AS df
ON df.Embarked = c.code

GROUP BY df.Embarked,
    df.Sex

""")

city Embarked     Sex    Avg_Age   Sum_Age  Min_Age  Max_Age  \
0       London     None    None        NaN       NaN      NaN      NaN   
1    Cherbourg        C  female  28.344262   1729.00     0.75     60.0   
2    Cherbourg        C    male  32.998841   2276.92     0.42     71.0   
3  Southampton        S  female  27.771505   5165.50     1.00     63.0   
4  Southampton        S    male  30.291440  11147.25     0.67     80.0   

   Total_passengers  Unique_tickets  
0                 0               0  
1                73              54  
2                95              88  
3               203             158  
4               441             394

<b>FULL JOIN</b><br>
FULL JOIN gives all rows from the from both tables, irrespective of common values

pandasql do not support FULL JOIN too. The output has both common and non-common elements from both the tables.

In [127]:
ps.sqldf("""

SELECT 
    c.city,
    df.Embarked,
    df.Sex,
    AVG(df.Age) AS Avg_Age,
    SUM(df.Age) AS Sum_Age,
    MIN(df.Age) AS Min_Age,
    MAX(df.Age) AS Max_Age,
    COUNT(df.PassengerId) AS Total_passengers,
    COUNT(DISTINCT df.Ticket) AS Unique_tickets
    
FROM cities AS c

FULL OUTER JOIN titanic_disaster_dataset AS df
ON df.Embarked = c.code

GROUP BY df.Embarked,
    df.Sex

""")

PandaSQLException: (sqlite3.OperationalError) RIGHT and FULL OUTER JOINs are not currently supported
[SQL: 

SELECT 
    c.city,
    df.Embarked,
    df.Sex,
    AVG(df.Age) AS Avg_Age,
    SUM(df.Age) AS Sum_Age,
    MIN(df.Age) AS Min_Age,
    MAX(df.Age) AS Max_Age,
    COUNT(df.PassengerId) AS Total_passengers,
    COUNT(DISTINCT df.Ticket) AS Unique_tickets
    
FROM cities AS c

FULL OUTER JOIN titanic_disaster_dataset AS df
ON df.Embarked = c.code

GROUP BY df.Embarked,
    df.Sex

]
(Background on this error at: http://sqlalche.me/e/e3q8)

<b>Union and Union All</b><br>
If you want to pull data from two datasets and stack one on top of another, we can use UNION.<br>
"UNION" statement gets all the distinct rows whereas "UNION ALL" simply stack the two tables one over another.

In [15]:
ps.sqldf("""

SELECT df.Embarked FROM titanic_disaster_dataset AS df
UNION
SELECT c.code FROM cities AS c

""")

Embarked
0     None
1        C
2        L
3        Q
4        S

##### Note that each SELECT statement within the UNION must have the same number of columns in the result sets.

In [16]:
ps.sqldf("""

SELECT DISTINCT df.Embarked FROM titanic_disaster_dataset AS df
UNION ALL
SELECT DISTINCT c.code FROM cities AS c

""")

Embarked
0        S
1        C
2        Q
3     None
4        S
5        C
6        L

<b>Minus or Except</b><br>
While "UNION" combines the datasets, "MINUS" or "EXCEPT" (only in some SQL dialects) gives the rows from the first dataset which are not present in the second one. pandasql uses "EXCEPT" keyword. Except "Q" and NULL, all values are present in "cities" table so only "Q" and Null values are pulled. 

In [17]:
ps.sqldf("""

SELECT DISTINCT df.Embarked FROM titanic_disaster_dataset AS df
EXCEPT
SELECT DISTINCT c.code FROM cities AS c

""")

Embarked
0     None
1        Q

Lets try vice versa for better understanding.<br> Only "L" is not present in "titanic_disaster_dataset" table when compared to "cities".

In [18]:
ps.sqldf("""

SELECT DISTINCT c.code FROM cities AS c
EXCEPT
SELECT DISTINCT df.Embarked FROM titanic_disaster_dataset AS df

""")

code
0    L

## Ranking in SQL

#### RANK(), DENSE_RANK() and ROW_NUMBERS() are three functions which help you to sort the data and give rank to each row.<br>
Generally these functions are used with OVER() function to sort the data based on some column before giving rank to each row and also to partition the data.

RANK() provides a rank to each row. If any n rows have the same value in the sorting column, then RANK() gives the same value to all of these rows. For the next row where the value is different than the previous n rows, the rank given will be previous rank plus n.<br>
Notice row no. 2 and 3, where the age values are the same(i.e. 71) and hence both rows have the same rank(i.e. 3). In row 4, the ranking begins from 5(i.e. previous rank 3 plus 2) and not 4.

In [21]:
ps.sqldf("""

SELECT 
    df.Name,
    df.Embarked,
    df.Sex,
    df.Age,
    RANK() OVER(ORDER BY df.Age DESC) AS Rank
FROM titanic_disaster_dataset AS df

""")

Name Embarked     Sex   Age  Rank
0        Barkworth, Mr. Algernon Henry Wilson        S    male  80.0     1
1                         Svensson, Mr. Johan        S    male  74.0     2
2                   Goldschmidt, Mr. George B        C    male  71.0     3
3                     Artagaveytia, Mr. Ramon        C    male  71.0     3
4                        Connors, Mr. Patrick        Q    male  70.5     5
..                                        ...      ...     ...   ...   ...
886                          Razi, Mr. Raihed        C    male   NaN   715
887         Sage, Miss. Dorothy Edith "Dolly"        S  female   NaN   715
888               van Melkebeke, Mr. Philemon        S    male   NaN   715
889                        Laleff, Mr. Kristo        S    male   NaN   715
890  Johnston, Miss. Catherine Helen "Carrie"        S  female   NaN   715

[891 rows x 5 columns]

RANK() will skip rank if same values found in the sorting order.<br>
To tackle this we can use DENSE_RANK() to ensure that no ranks are skipped.<br>
Take a look at the example below, where row no 4 has a rank 4 and not 5.

In [23]:
ps.sqldf("""

SELECT 
    df.Name,
    df.Embarked,
    df.Sex,
    df.Age,
    DENSE_RANK() OVER(ORDER BY df.Age DESC) AS Rank
FROM titanic_disaster_dataset AS df

""")

Name Embarked     Sex   Age  Rank
0        Barkworth, Mr. Algernon Henry Wilson        S    male  80.0     1
1                         Svensson, Mr. Johan        S    male  74.0     2
2                   Goldschmidt, Mr. George B        C    male  71.0     3
3                     Artagaveytia, Mr. Ramon        C    male  71.0     3
4                        Connors, Mr. Patrick        Q    male  70.5     4
..                                        ...      ...     ...   ...   ...
886                          Razi, Mr. Raihed        C    male   NaN    89
887         Sage, Miss. Dorothy Edith "Dolly"        S  female   NaN    89
888               van Melkebeke, Mr. Philemon        S    male   NaN    89
889                        Laleff, Mr. Kristo        S    male   NaN    89
890  Johnston, Miss. Catherine Helen "Carrie"        S  female   NaN    89

[891 rows x 5 columns]

Now, if suppose we do not want to repeat ranks and it should continue ranking on values. Here we can use ROW_NUMBER() which gives different rank to each row even if the sorted values are same.<br>
Notice below row no 2 and 3 which has differnt ranks even though Age is same.

In [24]:
ps.sqldf("""

SELECT 
    df.Name,
    df.Embarked,
    df.Sex,
    df.Age,
    ROW_NUMBER() OVER(ORDER BY df.Age DESC) AS Rank
FROM titanic_disaster_dataset AS df

""")

Name Embarked     Sex   Age  Rank
0        Barkworth, Mr. Algernon Henry Wilson        S    male  80.0     1
1                         Svensson, Mr. Johan        S    male  74.0     2
2                   Goldschmidt, Mr. George B        C    male  71.0     3
3                     Artagaveytia, Mr. Ramon        C    male  71.0     4
4                        Connors, Mr. Patrick        Q    male  70.5     5
..                                        ...      ...     ...   ...   ...
886                          Razi, Mr. Raihed        C    male   NaN   887
887         Sage, Miss. Dorothy Edith "Dolly"        S  female   NaN   888
888               van Melkebeke, Mr. Philemon        S    male   NaN   889
889                        Laleff, Mr. Kristo        S    male   NaN   890
890  Johnston, Miss. Catherine Helen "Carrie"        S  female   NaN   891

[891 rows x 5 columns]

We have used ORDER BY clause on above query but with single column. Can we use multiple columns and specify sorting order, the answer is YES.

In [26]:
ps.sqldf("""

SELECT 
    df.Name,
    df.Embarked,
    df.Sex,
    df.Age,
    df.Fare,
    RANK() OVER(ORDER BY df.Age DESC, df.Fare ASC) AS rnk
    
FROM titanic_disaster_dataset AS df

""")

Name Embarked     Sex   Age  \
0                 Barkworth, Mr. Algernon Henry Wilson        S    male  80.0   
1                                  Svensson, Mr. Johan        S    male  74.0   
2                            Goldschmidt, Mr. George B        C    male  71.0   
3                              Artagaveytia, Mr. Ramon        C    male  71.0   
4                                 Connors, Mr. Patrick        Q    male  70.5   
..                                                 ...      ...     ...   ...   
886                            Fleming, Miss. Margaret        C  female   NaN   
887  Frauenthal, Mrs. Henry William (Clara Heinshei...        S  female   NaN   
888     Spencer, Mrs. William Augustus (Marie Eugenie)        C  female   NaN   
889                                 Farthing, Mr. John        S    male   NaN   
890                                Robbins, Mr. Victor        C    male   NaN   

         Fare  rnk  
0     30.0000    1  
1      7.7750    2  
2     34.6542    3  
3     49.5042    4  
4      7.7500    5  
..        ...  ...  
886  110.8833  887  
887  133.6500  888  
888  146.5208  889  
889  221.7792  890  
890  227.5250  891  

[891 rows x 6 columns]

Previously i had mentioned that OVER() is used to sort the data based on some column before giving rank to each row and it can be used to partition the data too.<br>
Suppose you want to give separate ranks withtin specific levels of data, then you can use "PARTITION BY" in OVER() function. In the example below, for each combination of df.Embarked and df.Sex, we will get different sets of ranks. Refer row no 2 and 3, a different ranking is applied.

In [27]:
ps.sqldf("""

SELECT 
    df.Name,
    df.Embarked,
    df.Sex,
    df.Age,
    df.Fare,
    RANK() OVER(PARTITION BY df.Embarked, df.Sex 
                ORDER BY df.Age DESC, df.Fare DESC) AS rnk
    
FROM titanic_disaster_dataset AS df

""")

Name Embarked     Sex   Age  \
0           Stone, Mrs. George Nelson (Martha Evelyn)     None  female  62.0   
1                                 Icard, Miss. Amelie     None  female  38.0   
2    Warren, Mrs. Frank Manley (Anna Sophia Atkinson)        C  female  60.0   
3                                Lurette, Miss. Elise        C  female  58.0   
4       Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)        C  female  56.0   
..                                                ...      ...     ...   ...   
886                  Frost, Mr. Anthony Wood "Archie"        S    male   NaN   
887                     Parr, Mr. William Henry Marsh        S    male   NaN   
888                        Watson, Mr. Ennis Hastings        S    male   NaN   
889                              Knight, Mr. Robert J        S    male   NaN   
890                                  Fry, Mr. Richard        S    male   NaN   

         Fare  rnk  
0     80.0000    1  
1     80.0000    2  
2     75.2500    1  
3    146.5208    2  
4     83.1583    3  
..        ...  ...  
886    0.0000  434  
887    0.0000  434  
888    0.0000  434  
889    0.0000  434  
890    0.0000  434  

[891 rows x 6 columns]

<b>Lead and Lag</b><br>
LEAD() and LAG() can be used to get the nth previous or nth next value in a column, after sorting the data using OVER()<br>
In these functions the third parameter is returned in result when previous/next value is not found,. If you do not mention third parameter it will return NULL.

In [29]:
ps.sqldf("""

SELECT 
    df.Name,
    LEAD(df.Name, 1,0) OVER(ORDER BY df.Name) AS next_name,
    LEAD(df.Name, 2, 0) OVER(ORDER BY df.Name) AS second_next_name,
    LAG(df.Name, 1, 0) OVER(ORDER BY df.Name) AS previous_name,
    LAG(df.Name, 2, 0) OVER(ORDER BY df.Name) AS second_previous_name
    
FROM titanic_disaster_dataset AS df

""")

Name  \
0                      Abbing, Mr. Anthony   
1              Abbott, Mr. Rossmore Edward   
2         Abbott, Mrs. Stanton (Rosa Hunt)   
3                      Abelson, Mr. Samuel   
4    Abelson, Mrs. Samuel (Hannah Wizosky)   
..                                     ...   
886                de Mulder, Mr. Theodore   
887              de Pelsmaeker, Mr. Alfons   
888              del Carlo, Mr. Sebastiano   
889        van Billiard, Mr. Austin Blyler   
890            van Melkebeke, Mr. Philemon   

                                 next_name  \
0              Abbott, Mr. Rossmore Edward   
1         Abbott, Mrs. Stanton (Rosa Hunt)   
2                      Abelson, Mr. Samuel   
3    Abelson, Mrs. Samuel (Hannah Wizosky)   
4           Adahl, Mr. Mauritz Nils Martin   
..                                     ...   
886              de Pelsmaeker, Mr. Alfons   
887              del Carlo, Mr. Sebastiano   
888        van Billiard, Mr. Austin Blyler   
889            van Melkebeke, Mr. Philemon   
890                                      0   

                          second_next_name  \
0         Abbott, Mrs. Stanton (Rosa Hunt)   
1                      Abelson, Mr. Samuel   
2    Abelson, Mrs. Samuel (Hannah Wizosky)   
3           Adahl, Mr. Mauritz Nils Martin   
4                          Adams, Mr. John   
..                                     ...   
886              del Carlo, Mr. Sebastiano   
887        van Billiard, Mr. Austin Blyler   
888            van Melkebeke, Mr. Philemon   
889                                      0   
890                                      0   

                                    previous_name  \
0                                               0   
1                             Abbing, Mr. Anthony   
2                     Abbott, Mr. Rossmore Edward   
3                Abbott, Mrs. Stanton (Rosa Hunt)   
4                             Abelson, Mr. Samuel   
..                                            ...   
886  de Messemaeker, Mrs. Guillaume Joseph (Emma)   
887                       de Mulder, Mr. Theodore   
888                     de Pelsmaeker, Mr. Alfons   
889                     del Carlo, Mr. Sebastiano   
890               van Billiard, Mr. Austin Blyler   

                             second_previous_name  
0                                               0  
1                                               0  
2                             Abbing, Mr. Anthony  
3                     Abbott, Mr. Rossmore Edward  
4                Abbott, Mrs. Stanton (Rosa Hunt)  
..                                            ...  
886                            Zimmerman, Mr. Leo  
887  de Messemaeker, Mrs. Guillaume Joseph (Emma)  
888                       de Mulder, Mr. Theodore  
889                     de Pelsmaeker, Mr. Alfons  
890                     del Carlo, Mr. Sebastiano  

[891 rows x 5 columns]

Using "PARTITION BY" within OVER() function, you can partition the data and get LEAD() and LAG() of a column within partitions

In [49]:
ps.sqldf("""
SELECT df.Name, df.Embarked,
LEAD(df.Name, 1) OVER(PARTITION BY df.Embarked ORDER BY df.Name DESC) AS df_name
FROM titanic_disaster_dataset AS df
---ORDER BY df.Embarked
LIMIT 5
""")

Name Embarked                df_name
0  Stone, Mrs. George Nelson (Martha Evelyn)     None    Icard, Miss. Amelie
1                        Icard, Miss. Amelie     None                   None
2                  del Carlo, Mr. Sebastiano        C  Zabour, Miss. Thamine
3                      Zabour, Miss. Thamine        C   Zabour, Miss. Hileni
4                       Zabour, Miss. Hileni        C  Yousseff, Mr. Gerious

In [39]:
ps.sqldf("""

SELECT 
    df.Name,
    df.Embarked,
    LEAD(df.Name, 1,0) OVER(PARTITION BY df.Embarked ORDER BY df.Name DESC) AS next_name,
    LEAD(df.Name, 2, 0) OVER(PARTITION BY df.Embarked ORDER BY df.Name) AS second_next_name,
    LAG(df.Name, 1, 0) OVER(PARTITION BY df.Embarked ORDER BY df.Name) AS previous_name,
    LAG(df.Name, 2, 0) OVER(PARTITION BY df.Embarked ORDER BY df.Name) AS second_previous_name
    
FROM titanic_disaster_dataset AS df
LIMIT 15
""")

Name Embarked  \
0          Stone, Mrs. George Nelson (Martha Evelyn)     None   
1                                Icard, Miss. Amelie     None   
2                          del Carlo, Mr. Sebastiano        C   
3                              Zabour, Miss. Thamine        C   
4                               Zabour, Miss. Hileni        C   
5                              Yousseff, Mr. Gerious        C   
6                                  Yousif, Mr. Wazli        C   
7                               Youseff, Mr. Gerious        C   
8                           Young, Miss. Marie Grice        C   
9            Yasbeck, Mrs. Antoni (Selini Alexander)        C   
10                               Yasbeck, Mr. Antoni        C   
11                       Williams, Mr. Charles Duane        C   
12                         Widener, Mr. Harry Elkins        C   
13  Warren, Mrs. Frank Manley (Anna Sophia Atkinson)        C   
14                                  Ward, Miss. Anna        C   

                                           next_name  \
0                                Icard, Miss. Amelie   
1                                                  0   
2                              Zabour, Miss. Thamine   
3                               Zabour, Miss. Hileni   
4                              Yousseff, Mr. Gerious   
5                                  Yousif, Mr. Wazli   
6                               Youseff, Mr. Gerious   
7                           Young, Miss. Marie Grice   
8            Yasbeck, Mrs. Antoni (Selini Alexander)   
9                                Yasbeck, Mr. Antoni   
10                       Williams, Mr. Charles Duane   
11                         Widener, Mr. Harry Elkins   
12  Warren, Mrs. Frank Manley (Anna Sophia Atkinson)   
13                                  Ward, Miss. Anna   
14                          Uruchurtu, Don. Manuel E   

                           second_next_name  \
0                                         0   
1                                         0   
2                                         0   
3                                         0   
4                 del Carlo, Mr. Sebastiano   
5                     Zabour, Miss. Thamine   
6                      Zabour, Miss. Hileni   
7                     Yousseff, Mr. Gerious   
8                         Yousif, Mr. Wazli   
9                      Youseff, Mr. Gerious   
10                 Young, Miss. Marie Grice   
11  Yasbeck, Mrs. Antoni (Selini Alexander)   
12                      Yasbeck, Mr. Antoni   
13              Williams, Mr. Charles Duane   
14                Widener, Mr. Harry Elkins   

                                       previous_name  \
0                                Icard, Miss. Amelie   
1                                                  0   
2                              Zabour, Miss. Thamine   
3                               Zabour, Miss. Hileni   
4                              Yousseff, Mr. Gerious   
5                                  Yousif, Mr. Wazli   
6                               Youseff, Mr. Gerious   
7                           Young, Miss. Marie Grice   
8            Yasbeck, Mrs. Antoni (Selini Alexander)   
9                                Yasbeck, Mr. Antoni   
10                       Williams, Mr. Charles Duane   
11                         Widener, Mr. Harry Elkins   
12  Warren, Mrs. Frank Manley (Anna Sophia Atkinson)   
13                                  Ward, Miss. Anna   
14                          Uruchurtu, Don. Manuel E   

                                second_previous_name  
0                                                  0  
1                                                  0  
2                               Zabour, Miss. Hileni  
3                              Yousseff, Mr. Gerious  
4                                  Yousif, Mr. Wazli  
5                               Youseff, Mr. Gerious  
6                           Young, Miss. Marie Grice  
7            Yasbeck

<b>Subqueries</b><br>
Subqueries are small queries which are used within a bigger main SQL query. Subqueries store the output of a SQL temporarily and more SQL queries can be written on top of these.

In [52]:
ps.sqldf("""

SELECT df1.Embarked, df1.avg_age
    
FROM (SELECT a.Embarked, AVG(a.Age) AS avg_age 
      FROM titanic_disaster_dataset AS a
      GROUP BY a.Embarked) AS df1

WHERE df1.avg_age<30
    
""")

Embarked    avg_age
0        Q  28.089286
1        S  29.445397

Let's break the above query. Below output is for subquery. We are pulling data fro this output on the condition that avg_age should be less than 30. This is what we have received in above output.

In [51]:
ps.sqldf("""
SELECT a.Embarked, AVG(a.Age) AS avg_age 
      FROM titanic_disaster_dataset AS a
      GROUP BY a.Embarked
""")

Embarked    avg_age
0     None  50.000000
1        C  30.814769
2        Q  28.089286
3        S  29.445397

Let's look at one more example...<br>
We will pull average age on Embarked column wise<br>
then we will pull total passengers on Embarked column wise<br>
Further, we will combine these 2 results with JOIN to get common values

In [55]:
ps.sqldf("""
SELECT a.Embarked, AVG(a.Age) AS avg_age 
      FROM titanic_disaster_dataset AS a
      GROUP BY a.Embarked    
""")

Embarked    avg_age
0     None  50.000000
1        C  30.814769
2        Q  28.089286
3        S  29.445397

In [56]:
ps.sqldf("""

SELECT b.Embarked, COUNT(b.PassengerID) AS tot_passengers
     FROM titanic_disaster_dataset AS b
     GROUP BY b.Embarked

""")

Embarked  tot_passengers
0     None               2
1        C             168
2        Q              77
3        S             644

In [54]:
ps.sqldf("""

SELECT df1.Embarked, df1.avg_age, df2.tot_passengers
FROM (SELECT a.Embarked, AVG(a.Age) AS avg_age 
      FROM titanic_disaster_dataset AS a
      GROUP BY a.Embarked) AS df1

INNER JOIN

    (SELECT b.Embarked, COUNT(b.PassengerID) AS tot_passengers
     FROM titanic_disaster_dataset AS b
     GROUP BY b.Embarked) AS df2

ON df1.Embarked = df2.Embarked
    
""")

Embarked    avg_age  tot_passengers
0        C  30.814769             168
1        Q  28.089286              77
2        S  29.445397             644

<b> Nested queries </b><br>
We can use nested subqueries inside subquery and also use it to filter the data.<br>
We are running subqueries separetly to check how the query come up with the final output.

In [65]:
ps.sqldf("""
SELECT b.* FROM titanic_disaster_dataset AS b
WHERE b.Embarked IN 
       (SELECT DISTINCT df1.Embarked
        FROM (SELECT a.Embarked, AVG(a.Age) AS avg_age 
              FROM titanic_disaster_dataset AS a
              GROUP BY a.Embarked) AS df1
        WHERE df1.avg_age>30)
LIMIT 5
""")

PassengerId  Survived  Pclass  \
0            2         1       1   
1           10         1       2   
2           20         1       3   
3           27         0       3   
4           31         0       1   

                                                Name     Sex   Age  SibSp  \
0  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
1                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
2                            Masselmani, Mrs. Fatima  female   NaN      0   
3                            Emir, Mr. Farred Chehab    male   NaN      0   
4                           Uruchurtu, Don. Manuel E    male  40.0      0   

   Parch    Ticket     Fare Cabin Embarked  
0      0  PC 17599  71.2833   C85        C  
1      0    237736  30.0708  None        C  
2      0      2649   7.2250  None        C  
3      0      2631   7.2250  None        C  
4      0  PC 17601  27.7208  None        C

In [61]:
ps.sqldf("""
SELECT DISTINCT df1.Embarked
        FROM (SELECT a.Embarked, AVG(a.Age) AS avg_age 
              FROM titanic_disaster_dataset AS a
              GROUP BY a.Embarked) AS df1
        WHERE df1.avg_age>30
        """)

Embarked
0     None
1        C

So below we have all values of Embarked column with average age. We pull avg_age on condition that it should be > 30 which gave only "C" as output. So our final output has rows which has Emabrked with values "C".

In [60]:
ps.sqldf("""
SELECT a.Embarked, AVG(a.Age) AS avg_age 
              FROM titanic_disaster_dataset AS a
              GROUP BY a.Embarked
              """)

Embarked    avg_age
0     None  50.000000
1        C  30.814769
2        Q  28.089286
3        S  29.445397

<b>Typecast</b><br>
You can use CAST() function to change the data type of columns in a dataset. Here are some example - 

In [67]:
ps.sqldf("""

SELECT 
    df.Name,
    df.Fare,
    CAST(df.Fare AS INT) AS integer_fare,
    CAST(CAST(df.Fare AS INT) AS FLOAT) AS float_fare
FROM titanic_disaster_dataset AS df
LIMIT 5
    
""")

Name     Fare  integer_fare  \
0                            Braund, Mr. Owen Harris   7.2500             7   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  71.2833            71   
2                             Heikkinen, Miss. Laina   7.9250             7   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  53.1000            53   
4                           Allen, Mr. William Henry   8.0500             8   

   float_fare  
0         7.0  
1        71.0  
2         7.0  
3        53.0  
4         8.0